### PR TITLE
Fetch cipher-wrap after loading providers.

### PR DESCRIPTION
**Pull request by richsalz**
Should this also handle the hardcoded wrap options as well?
```
  {"aes192-wrap", OPT_AES192_WRAP, '-', "Use AES192 to wrap key"},
    {"aes256-wrap", OPT_AES256_WRAP, '-', "Use AES256 to wrap key"},
# ifndef OPENSSL_NO_DES
    {"des3-wrap", OPT_3DES_WRAP, '-', "Use 3DES-EDE to wrap key"},
# endif
    {"wrap", OPT_WRAP, 's', "Any wrap cipher to wrap key"},

```

If so, what are the official names for the wrapping ciphers?  And should the `no_des` ifdef be removed? From here?  From all apps?

### CODE REVIEW

File: apps/lib/opt.c
Comment (levitte) [inline apps/lib/opt.c:361]: Unfortunately, this introduces a leak.

The quick hack I made was this:

``` diff
diff --git a/apps/lib/opt.c b/apps/lib/opt.c
index 8cc520daec..becf46f11b 100644
--- a/apps/lib/opt.c
+++ b/apps/lib/opt.c
@@ -353,9 +353,12 @@ void print_format_error(int format, unsigned long flags)
 }
 
 /* Parse a cipher name, put it in *EVP_CIPHER; return 0 on failure, else 1. */
-int opt_cipher(const char *name, const EVP_CIPHER **cipherp)
+int opt_cipher(const char *name,
+               const EVP_CIPHER **cipherp, EVP_CIPHER **fetched_cipherp)
 {
-    *cipherp = EVP_get_cipherbyname(name);
+    if ((*cipherp = EVP_get_cipherbyname(name)) != NULL
+        || (*cipherp = *fetched_cipherp =
+            EVP_CIPHER_fetch(NULL, name, NULL)) != NULL)
     if (*cipherp != NULL)
         return 1;
     opt_printf_stderr("%s: Unknown cipher: %s\n", prog, name);
@@ -365,9 +368,11 @@ int opt_cipher(const char *name, const EVP_CIPHER **cipherp)
 /*
  * Parse message digest name, put it in *EVP_MD; return 0 on failure, else 1.
  */
-int opt_md(const char *name, const EVP_MD **mdp)
+int opt_md(const char *name, const EVP_MD **mdp, EVP_MD **fetched_mdp)
 {
-    *mdp = EVP_get_digestbyname(name);
+    if ((*mdp = EVP_get_digestbyname(name)) != NULL
+        || (*mdp = *fetched_mdp =
+            EVP_MD_fetch(NULL, name, NULL)) != NULL)
     if (*mdp != NULL)
         return 1;
     opt_printf_stderr("%s: Unknown option or message digest: %s\n", prog,
```

Of course, it means that every caller needs to adapt.  That's not so bad, here's the result in `apps/enc.c`:

``` diff
diff --git a/apps/enc.c b/apps/enc.c
index 9982337c01..dfb5c239ad 100644
--- a/apps/enc.c
+++ b/apps/enc.c
@@ -109,8 +110,8 @@ int enc_main(int argc, char **argv)
     BIO *in = NULL, *out = NULL, *b64 = NULL, *benc = NULL, *rbio =
         NULL, *wbio = NULL;
     EVP_CIPHER_CTX *ctx = NULL;
-    const EVP_CIPHER *cipher = NULL;
-    const EVP_MD *dgst = NULL;
+    const EVP_CIPHER *cipher = NULL; EVP_CIPHER *fetched_cipher = NULL;
+    const EVP_MD *dgst = NULL; EVP_MD *fetched_dgst = NULL;
     const char *digestname = NULL;
     char *hkey = NULL, *hiv = NULL, *hsalt = NULL, *p;
     char *infile = NULL, *outfile = NULL, *prog;
@@ -297,7 +303,7 @@ int enc_main(int argc, char **argv)
 
     /* Get the cipher name, either from progname (if set) or flag. */
     if (ciphername != NULL) {
-        if (!opt_cipher(ciphername, &cipher))
+        if (!opt_cipher(ciphername, &cipher, &fetched_cipher))
             goto opthelp;
     }
     if (cipher && EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER) {
@@ -309,7 +315,7 @@ int enc_main(int argc, char **argv)
         goto end;
     }
     if (digestname != NULL) {
-        if (!opt_md(digestname, &dgst))
+        if (!opt_md(digestname, &dgst, &fetched_dgst))
             goto opthelp;
     }
     if (dgst == NULL)
@@ -625,6 +634,8 @@ int enc_main(int argc, char **argv)
         BIO_printf(bio_err, "bytes written: %8ju\n", BIO_number_written(out));
     }
  end:
+    EVP_CIPHER_free(fetched_cipher);
+    EVP_MD_free(fetched_dgst);
     ERR_print_errors(bio_err);
     OPENSSL_free(strbuf);
     OPENSSL_free(buff);
```

**Comment by richsalz** [comment]
> Unfortunately, this introduces a leak.

And cleaning it up becomes a REAL pain,  const/non-const, etc.  It would be *great* if the EVP types had a built-in flag so that you could always call free and it would be ignore if you were returning a built-in global.  Sigh.

**Comment by paulidale** [comment]
I believe that the proper solution for this is Rich wrote.

**Comment by levitte** [comment]
> I believe that the proper solution for this is Rich wrote.

I've considered that before, but quite frankly, ewwwww.  Basically, what you're suggesting is to have application deliberately break constness somewhere...  I can see how that will fly

**Comment by richsalz** [comment]
I am willing to do the add-flag thing if the OTC is in favor.

I don't like breaking const.  More importantly, tho, having to do things like you show above in https://github.com/openssl/openssl/pull/14182#discussion_r575867943 is even worse.  But this is just a matter of style and taste, and the OMC needs to decide because it is holding up other work (like this PR).

**Comment by levitte** [comment]
You gotta remember that this "fix" is also transitional.  In some future, fetching will be the only recourse, so the question is how far you want to go to be able to "nicely" support legacy stuff...

**Comment by richsalz** [comment]
If the fix is temporary, then the cleaner thing to do is add the flag.

**Comment by slontis** [comment]
> If so, what are the official names for the wrapping ciphers? And should the `no_des` ifdef be removed? From here? From all apps?

SP800-38F says it is Key Wrap

**Comment by slontis** [comment]
> Basically, what you're suggesting is to have application deliberately break constness somewhere... I can see how that will fly

So in this case opt_md for example should not return a const EVP_MD*.

const EVP_MD should have flag that is off.. The fetch sets the flag for the dynamic case.. 

Seems like a reasonable idea to me.

**Comment by levitte** [comment]
> So in this case opt_md for example should not return a const EVP_MD*.

So, does that mean that we change `EVP_get_digestbyname()` to return a non-const `EVP_MD *`?
Or does that mean that we un-const its return value in `opt_md()`?
Also, is that what you'd recommend other applications to do?

**Comment by slontis** [comment]
> > So in this case opt_md for example should not return a const EVP_MD*.
> 
> So, does that mean that we change `EVP_get_digestbyname()` to return a non-const `EVP_MD *`?
> Or does that mean that we un-const its return value in `opt_md()`?
> Also, is that what you'd recommend other applications to do?

Ahh ok, changing EVP_get_digestbyname() to non const sounds like a bad idea..
If opt_md() is making a choice between a const/ non const then it would need to cast.
We already had to change the code to add the fetch, so that is the point where you need to do the cast if you need both.
This would then propogate to anywhere where you had to do the selection and then free (which we have in quite a few places already).

**Comment by richsalz** [comment]
Mostly agree with @slontis's comment at https://github.com/openssl/openssl/pull/14182#issuecomment-778870130 but I'd make the local variable non-const.

As for what advice to give developers, that's up to the project to decide.  If it doesn't like my "don't free me" flag in the structure, it'll have to decide what to do instead.

**Comment by richsalz** [comment]
> SP800-38F says it is Key Wrap

Yeah, but I was thinking of 'openssl official' as in https://github.com/openssl/openssl/blob/master/crypto/evp/e_aes.c#L3625

**Comment by paulidale** [comment]
I don't understand the talk about const/non-const.  The flag will be fixed in the structures EVP_get_digestbyname() returns, they are not going to be modified.  Sure a cast will be required when freeing them but the underlying memory won't be altered.

Adding a flag and insisting on some casts is a small price compared to having additional arguments to functions or carrying two pointers everywhere.  We've already got cases where we've a _fetched_ pointer and the _real_ pointer, they are ugly.

**Comment by richsalz** [comment]
> I don't understand the talk about const/non-const. 

I think you're looking for more than there is.  IT's all about sometimes you get a const pointer via `EVP_xxx()` or `EVP_get...byname()` or sometimes you get a pointer via a `fetch` call.  And as you point out, when you have to do a mix of things (as in the apps and other code), it's yucky.

So far @levitte has been arguing against cast-away-const, and others feel the pain of the current code isn't worth it.

**Comment by paulidale** [comment]
We could hide the cast away of const inside the free call.  I don't like this as much as making folks cast it away in their code because we will be changing to the non-const version eventually.

People who grab the algorithm via `EVP_xxx()` aren't going to be freeing it.  Those who update their code to fetch will be.  The majority of the cases where both forms are necessary will be in our libraries.  Let's make it easy for us right now....

**Comment by levitte** [comment]
So @paulidale, it sounds like you're planning on something like this:

``` C
EVP_MD_free(const EVP_MD *);
```

From an API design point of view, that is *weird*.
If we're doing this, I would say that it's wiser to do it like this:

``` diff
diff --git a/crypto/evp/digest.c b/crypto/evp/digest.c
index 3dfcfcda8e..1696d3fb11 100644
--- a/crypto/evp/digest.c
+++ b/crypto/evp/digest.c
@@ -1017,7 +1017,7 @@ void EVP_MD_free(EVP_MD *md)
 {
     int i;
 
-    if (md == NULL)
+    if (md == NULL || md->prov == NULL)
         return;
 
     CRYPTO_DOWN_REF(&md->refcnt, &i, md->lock);
diff --git a/crypto/evp/names.c b/crypto/evp/names.c
index cb59813857..1c8c84f741 100644
--- a/crypto/evp/names.c
+++ b/crypto/evp/names.c
@@ -105,29 +105,29 @@ const EVP_CIPHER *evp_get_cipherbyname_ex(OSSL_LIB_CTX *libctx,
 
 static void digest_from_name(const char *name, void *data)
 {
-    const EVP_MD **md = data;
+    EVP_MD **md = data;
 
     if (*md != NULL)
         return;
 
-    *md = (const EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
+    *md = (EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
 }
 
-const EVP_MD *EVP_get_digestbyname(const char *name)
+EVP_MD *EVP_get_digestbyname(const char *name)
 {
     return evp_get_digestbyname_ex(NULL, name);
 }
 
-const EVP_MD *evp_get_digestbyname_ex(OSSL_LIB_CTX *libctx, const char *name)
+EVP_MD *evp_get_digestbyname_ex(OSSL_LIB_CTX *libctx, const char *name)
 {
-    const EVP_MD *dp;
+    EVP_MD *dp;
     OSSL_NAMEMAP *namemap;
     int id;
 
     if (!OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_DIGESTS, NULL))
         return NULL;
 
-    dp = (const EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
+    dp = (EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);
 
     if (dp != NULL)
         return dp;
diff --git a/include/openssl/evp.h b/include/openssl/evp.h
index 5f9de9d8b9..7658a90c1a 100644
--- a/include/openssl/evp.h
+++ b/include/openssl/evp.h
@@ -1100,7 +1100,7 @@ int EVP_add_cipher(const EVP_CIPHER *cipher);
 int EVP_add_digest(const EVP_MD *digest);
 
 const EVP_CIPHER *EVP_get_cipherbyname(const char *name);
-const EVP_MD *EVP_get_digestbyname(const char *name);
+EVP_MD *EVP_get_digestbyname(const char *name);
 
 void EVP_CIPHER_do_all(void (*fn) (const EVP_CIPHER *ciph,
                                    const char *from, const char *to, void *x),
```

**Comment by levitte** [comment]
Do note the little detail that we don't need an extra flag, the `prov` pointer is already there to save the day...

**Comment by paulidale** [comment]
> Do note the little detail that we don't need an extra flag, the `prov` pointer is already there to save the day...

Even better :)

**Comment by beldmit** [comment]
So this snipped of code (https://github.com/openssl/openssl/blob/9dc9c7f2d7f25e100dd7d80e9bc51e205033cd8c/apps/speed.c#L1498-L1528) and corresponding `free` operations (https://github.com/openssl/openssl/blob/9dc9c7f2d7f25e100dd7d80e9bc51e205033cd8c/apps/speed.c#L4137-L4140) can be simplified?

**Comment by levitte** [comment]
BTW, I believe that EVP_XXX_fetch() should be a fallback after EVP_get_yyyyyybyname() fails, not the other way around.
The reason is that if an engine is loaded and implements an alternative for an algorithm that we have in our default provider, the engine code will never be used.  That will be a surprise.

**Comment by beldmit** [comment]
Fair point, but almost everywhere in code I remember that we try to fetch first...

**Comment by levitte** [comment]
> Fair point, but almost everywhere in code I remember that we try to fetch first...

Not quite...  there are a few places where, if we don't have prior knowledge, we have a look at whatever's available from engines *first*.  I haven't looked much at what happens with speed, so I hadn't noticed that it does it the other way around.  It would be interesting to know what happens if we try speed with, say, the padlock engine...

**Comment by levitte** [comment]
Speaking of what we're currently talking about, I expect to see that fixed in a *separate commit*, as that has nothing to do with cipher-wrap per se.  It should be marked with 'Fixes #14178' and 'Fixes #14179'.  That may actually make it suitable to be a separate PR.

**Comment by richsalz** [comment]
My plan would be to add the flag and change `opt_cipher` and `opt_md` to not return `const` values, and then always call the `EVP_xxx_free` function.

**Comment by levitte** [comment]
> My plan would be to add the flag

Did you catch on that it's not necessary?  Why add redundant information?

**Comment by t8m** [comment]
@levitte Could you please submit the patch from the comment above 
as a PR? As that change would be useful for my hacking of speed command as well.

**Comment by levitte** [comment]
Sure

**Comment by levitte** [comment]
It turns out that @richsalz has a point with a flag.  an EVP_MD is not only created with EVP_MD_fetch(), but also with EVP_MD_meth_new().

**Comment by levitte** [comment]
I'll take the rest of the discussion in #14178 and #14179, this PR has been derailed enough as it is.

**Comment by richsalz** [comment]
I have a PR coming that cleans up and standardizes all the `EVP_xxx_free` routines.  It looks like they all work as-is, except for `EVP_MD_free` which needs the flag and is what I had been looking at. :)

**Comment by levitte** [comment]
@richsalz please see what I wrote in #14179.

**Comment by richsalz** [comment]
This is on-hold until #14194 gets merged and then I'll either update it or abandon it.

**Comment by levitte** [comment]
> This is on-hold...

Er, ok, so the original intention with this PR, to have cipher-wrap fetched after providers are loaded, has somehow lost its value?

**Comment by richsalz** [comment]
I am only inclined to rebase once, and since there's other work happening in the apps I'm waiting.

**Comment by levitte** [comment]
Ah, sorry, quoted the wrong thing.  It was the mention of abandoning that surprised me.

**Comment by richsalz** [comment]
>  It was the mention of abandoning that surprised me.

Just trying to be completely realistic.

**Comment by richsalz** [comment]
Refreshing this since the OTC decided on an approach and #14193 was approved and merged.  Most of the discussion here can be ignored now. There are still a couple of open questions:
- Should the hardcoded cipher-wrap options be handled this way also? https://github.com/openssl/openssl/pull/14182#issue-573141714
- What are the "official" names (to be used for lookup/fetch) of those cipher-wraps?
- Should we remove the impact of the `no_des` flag since leaving it out of openssl still means another provider can do it.

**Comment by richsalz** [comment]
Removed "WIP" and made it a draft, since that seems to be the thing to do these days. Let me know if not.

File: apps/lib/opt.c
Comment (slontis) [inline apps/lib/opt.c:362]: Not sure I understand the order here.. Didnt we agree on the reverse order?
See https://github.com/openssl/openssl/issues/14198

File: apps/lib/opt.c
Comment (slontis) [inline apps/lib/opt.c:377]: Similar comment for here.

File: apps/lib/opt.c
Comment (richsalz) [reply apps/lib/opt.c:362]: We did, but it breaks tests, at least because some output changed.  Waiting for the "official names" PR to get merged and then we'll look at this.

Comment (t8m) [reply apps/lib/opt.c:362]: I've reopened #14198 so we do not forget.

**Comment by richsalz** [comment]
Waiting on answers from the OMC https://github.com/openssl/openssl/pull/14182#issuecomment-821999931

**Comment by t8m** [comment]
> Refreshing this since the OTC decided on an approach and #14193 was approved and merged. Most of the discussion here can be ignored now. There are still a couple of open questions:
> 
>     * Should the hardcoded cipher-wrap options be handled this way also? [#14182 (comment)](https://github.com/openssl/openssl/pull/14182#issue-573141714)
> 
>     * What are the "official" names (to be used for lookup/fetch) of those cipher-wraps?
> 
>     * Should we remove the impact of the `no_des` flag since leaving it out of openssl still means another provider can do it.

I do not think there is anything controversial here that would require OMC vote.

Answers IMO are:
1. yes
2. please look at the default provider - the first name of the algorithm is supposed to be the "canonical" name as we see it. Of course opinion of other provider authors might differ and that's the reason we allow multiple names. However they should at least include our "canonical" name as an alias if they want to be compatible
3. yes, if you want to, however I do not think this is a blocker for 3.0 at all

**Comment by richsalz** [comment]
This now depends on #15011

File: apps/cms.c
Comment (levitte) [inline apps/cms.c:705]: I would actually reduce all the specific wrap cases like this:

``` C
        case OPT_3DES_WRAP:
        case OPT_AES128_WRAP:
        case OPT_AES192_WRAP:
        case OPT_AES256_WRAP:
            wrapname = opt_flag() + 1;
            break;
```

The option strings happen to be perfectly legitimate names already, so...
With this, no need for first(), no need to move prov/names.h, ...

**Review by levitte — CHANGES_REQUESTED** [review]

File: apps/cms.c
Comment (t-j-h) [reply apps/cms.c:23]: No! The apps should **never** use internal header files and if you find your need to use an internal header file then we have the api **wrong** and it needs fixing.

And yes I know we have some "internal" headers in use currently that need fixing (socket,h and o_dir.h usage should not be internal nor should nelem.h and that is what is currently there).

Comment (richsalz) [reply apps/cms.c:23]: I am not thrilled with moving the file to a public place.  On the other hand, I don't think there's any alternative. How is the app/cms supposed to know what the official name is for aes128-wrapping function? Also, how are third-party crypto providers supposed to know what names they need to implement?

File: apps/cms.c
Comment (richsalz) [reply apps/cms.c:705]: That's clever!  Does fetch do strcasecmp or strcmp?
```
#define PROV_NAMES_AES_128_WRAP "AES-128-WRAP:id-aes128-wrap:AES128-WRAP:2.16.840.1.101.3.4.1.5"
```

Comment (levitte) [reply apps/cms.c:705]: strcasecmp

Comment (richsalz) [reply apps/cms.c:705]: > strcasecmp

Okay, that's great.  I'll do what you suggested.  I will also change `test/recipes/30-test_evp_data/evpciph_aes_wrap.txt` to have an uppercase and lowercase name.

File: apps/cms.c
Comment (richsalz) [reply apps/cms.c]: good thing we don't need to `OPENSSL_strlcpy`  any more :)

**Comment by richsalz** [comment]
Rebased against master and updated commit pushed.

File: apps/cms.c
Comment (levitte) [reply apps/cms.c]: The apps don't need to know *the* official name, they need to know *one of the names*.

The same goes for providers...  they really don't need to know any other name than they are willing to declare, i.e. the providers can be said to have an authority.  However, if one of their algorithms have at least one common name with an implementation from another providers, the union of all those names are becoming aliases for each other.  That's how the namemaps in libcrypto operate.

Most algorithms have some sort of official name, among others given by the inventor of the algorithm, but also other names, like the OID names given by NIST.  *We* don't really need to dictate this...  we do pre-populate our namemaps with some algorithm names known by libcrypto, though, and the rest is filled in as our providers, or other providers are loaded.

Comment (richsalz) [reply apps/cms.c]: Let's move this dicussion to #15011.  Marking this as reolved.  Please dup your comment there and the team can decide.

**Review by t8m — APPROVED** [review]

File: test/recipes/30-test_evp_data/evpciph_aes_wrap.txt
Comment (slontis) [inline test/recipes/30-test_evp_data/evpciph_aes_wrap.txt:10]: These were here deliberately to test the aliases..

Comment (t8m) [reply test/recipes/30-test_evp_data/evpciph_aes_wrap.txt:10]: So, @richsalz please keep the aliases but add a comment so it is not reverted again.

Comment (richsalz) [reply test/recipes/30-test_evp_data/evpciph_aes_wrap.txt]: Added comment, will copy tests with different case names and comment those.  Updated commit pushed.

**Review by t8m — APPROVED** [review]
still OK

**Review by DDvO — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged to master. Thank you for the contribution.

**Pull request closed without merging**
